compiler/rustc_hir_typeck/src/fn_ctxt/adjust_fulfillment_errors.rs

```diff
@@ -67,6 +74,33 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             })
         };
 
+        // Account for enum variant constructors, where the type param corresponds to the enum
```

Comment:
This logic is pretty different from all the other parameter searching logic below. I'm pretty confused by what it's trying to achieve -- can you explain at a high level what cases it's trying to account for, and why it's correct?

Reply:
It is to properly handle this case: https://github.com/rust-lang/rust/pull/115219/commits/9e4f0ac5707d176c501211d6b139fd8024c54ab2#diff-6ab1aef262db83c1d2b5744088ab0c1466ca205d77282627a17a16a5e41bfcf5

Reply:
#115221 should solve this more cleanly.